[Feature] Support zstd compression for data page (#10090)

srlch opened the pull request:
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #10090

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

Problem:
Currently, data pages use LZ4 for page compression by default, which is less flexible. This issue supports configurable page compression algorithms for page compression.

Solution:
Add the compression keyword to the properties of the CREATE TABLE statement to configure the compression algorithm for the table.

Performance:

test dataset:https://datasets.clickhouse.com/hits_compatible/hits.tsv.gz
compared to lz4:
19% higher compression ratio.
The query speed is increased by an average of 30% to 40% without OS page cache.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=10097) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=10097) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=10097) it.</sub>

imay commented on fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:
```diff
@@ -427,6 +430,29 @@ public static TStorageFormat analyzeStorageFormat(Map<String, String> properties
+            return TCompressionType.LZ4_FRAME;
+        }
+
+        if (compressionType.equalsIgnoreCase("lz4")) {
```
you can use CompressionUtils directly.

srlch replied:
OK, but in fact, not all compression algorithm in TCompressionType can be used for data pages by now. Comparison between different compression algorithm is unavoidable. This must be done in PropertyAnalyzer or in CompressionUtils. Of course, we can do it in CompressionUtils.

imay replied:
compressionutils is a common function, if storage does not supported some compression type, it is the storage to do the compression check. But storage can use TCompressionType after CompressionUtils converting.

srlch replied:
I think it is a good idea to do converting/checking in CompressionUtils before the storage use the compression type. Because, if the user exec the CREATE TABLE statement with a unsupported compression type, we should throw the error in analyzeCompressionType immediately.

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[707, 708]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2065]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

imay commented on be/src/storage/metadata_util.cpp:
```diff
@@ -223,6 +223,23 @@ Status convert_t_schema_to_pb_schema(const TTabletSchema& tablet_schema, uint32_
+        schema->set_compression_type(ZSTD);
+        break;
+    default:
+        CHECK(false) << "unexpected compression type " << compression_type;
```
NOT use CHECK, this will cause the process crash. Should return error for this per request problem

srlch replied:
OK,  I will fix it.

imay commented on gensrc/proto/tablet_schema.proto:
```diff
@@ -66,6 +67,7 @@ message TabletSchemaPB {
     optional uint32 next_column_unique_id = 7; // OLAPHeaderMessage.next_column_unique_id
     optional bool DEPRECATED_is_in_memory = 8 [default=false];
     optional int64 deprecated_id = 9; // deprecated
+    optional CompressionTypePB compression_type = 10 [default=LZ4_FRAME];
```
what's the difference between compression_type and compress_kind

srlch replied:
compress_kind is a deprecated variable for the storgae engine.

imay commented on be/src/storage/lake/tablet_manager.cpp:
```diff
@@ -181,18 +181,19 @@ Status TabletManager::create_tablet(const TCreateTabletReq& req) {
-                mutable_new_schema, next_unique_id, col_idx_to_unique_id, tablet_metadata_pb.mutable_schema()));
+        RETURN_IF_ERROR(
+                starrocks::convert_t_schema_to_pb_schema(mutable_new_schema, next_unique_id, col_idx_to_unique_id,
+                                                         tablet_metadata_pb.mutable_schema(), req.compression_type));
```
You should check whether compression_type is set in the `req`.
because this field is not set in the request from old version FE,

srlch replied:
The check of req and the error return of switch case in convert_t_schema_to_pb_schema can be considered as the same thing, and I will fix it together.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -1844,6 +1845,21 @@ public TStorageFormat getStorageFormat() {
         return tableProperty.getStorageFormat();
     }
 
+    public void setCompressionType(TCompressionType compressionType) {
```
Does the compressType can be modified using alter table?

srlch replied:
alter table is currently not supported.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/common/util/CompressionUtils.java:
```diff
@@ -25,4 +25,22 @@ public class CompressionUtils {
+                   || compressionType == TCompressionType.ZLIB
+                   || compressionType == TCompressionType.ZSTD) {
+            return compressionType;
+        } else {
```
If the compression type is not supported, we should return an exception or error to give an obvious hint

srlch replied:
Yes, we do. The exception or error will be returned in PropertyAnalyzer which call this function.

chaoyli commented on be/src/storage/metadata_util.cpp:
```diff
@@ -223,6 +223,23 @@ Status convert_t_schema_to_pb_schema(const TTabletSchema& tablet_schema, uint32_
     schema->set_compress_kind(COMPRESS_LZ4);
 
+    switch (compression_type) {
+    case TCompressionType::LZ4_FRAME:
```
```suggestion
    case TCompressionType::LZ4_FRAME:
    case TCompressionType::LZ4:
            schema->set_compression_type(LZ4);
```

chaoyli commented on be/src/storage/metadata_util.cpp:
```diff
@@ -223,6 +223,23 @@ Status convert_t_schema_to_pb_schema(const TTabletSchema& tablet_schema, uint32_
+    case TCompressionType::ZSTD:
+        schema->set_compression_type(ZSTD);
+        break;
+    default:
```
CHECK will crash. Use LOG(WARNING) and return a error instead.

srlch replied:
ok, I will fix it.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:
```diff
@@ -427,6 +430,23 @@ public static TStorageFormat analyzeStorageFormat(Map<String, String> properties
 
+    // analyzeCompressionType will parse the compression type from properties
+    public static TCompressionType analyzeCompressionType(Map<String, String> properties) throws AnalysisException {
+        String compressionType;
```
```You can first return the default operation.
        if (properties == null || !properties.containsKey(PROPERTIES_COMPRESSION)) {
            return TCompressionType.LZ4_FRAME;
        }
```

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:
```diff
@@ -427,6 +430,23 @@ public static TStorageFormat analyzeStorageFormat(Map<String, String> properties
+    public static TCompressionType analyzeCompressionType(Map<String, String> properties) throws AnalysisException {
+        String compressionType;
+        if (properties != null && properties.containsKey(PROPERTIES_COMPRESSION)) {
+            compressionType = properties.get(PROPERTIES_COMPRESSION);
```
Does it necessary to remove the PROPERTIES_COMPRESSION from properties?

srlch replied:
In PropertyAnalyzer.java, any other attributes in properties will be remove after analyzing. I think we should be consistent with the other attributes.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/common/util/CompressionUtils.java:
```diff
@@ -25,4 +25,22 @@ public class CompressionUtils {
+
+    // Return TCompressionType according to input name for data page compression.
+    // Return null if input name is an invalid compression type for data page compression.
+    public static TCompressionType findPageTCompressionByName(String name) {
```
getCompressTypeByName is OK. FE does not know the page concept.

srlch replied:
Yes, but the compression check must be done in fe as we can return the error in analyzeCompressionType immediately if we get a unsupported compression type. So I think it is necessary to do the check in fe?

chaoyli replied:
My meaning is the function name can be changed. `getCompressTypeByName` may be better

srlch commented on gensrc/thrift/AgentService.thrift:
```diff
@@ -76,6 +76,7 @@ struct TCreateTabletReq {
     13: optional TStorageFormat storage_format
     14: optional TTabletType tablet_type
     15: optional bool enable_persistent_index
+    16: optional Types.TCompressionType compression_type = Types.TCompressionType.LZ4_FRAME
```
LZ4_FRAME support mutiple slice input，so I think all LZ4 can be replaced by LZ4_FRAME. Maybe we can remove lz4 in the future.

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[707, 708]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2065]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[707, 708]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2063]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[707, 708]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2063]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[708, 709]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2068]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

srlch committed:
[Feature] Support zstd compression for data page (#10090)

Problem:
data pages use LZ4 for page compression by default, which is less flexible.
This issue supports configurable page compression algorithms for page compression.

Solution:
Add the compression keyword to the properties of the CREATE TABLE statement to configure the compression algorithm for the table.

Performance:

test dataset:https://datasets.clickhouse.com/hits_compatible/hits.tsv.gz
compared to lz4:
19% higher compression ratio.
The query speed is increased by an average of 30% to 40% without OS page cache.

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[708, 709]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2068]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 54 (87.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|2|00.00%|[795, 796]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|2|00.00%|[708, 709]|
|:large_blue_circle:|com/starrocks/common/util/CompressionUtils.java|5|6|83.33%|[43]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|6|7|85.71%|[2068]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|8|87.50%|[1850]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/CreateReplicaTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|1|1|100.00%|[]|

jaogoy commented on be/src/storage/lake/tablet_manager.cpp:
```diff
@@ -191,17 +191,18 @@ Status TabletManager::create_tablet(const TCreateTabletReq& req) {
         RETURN_IF_ERROR(starrocks::convert_t_schema_to_pb_schema(
-                mutable_new_schema, next_unique_id, col_idx_to_unique_id, tablet_metadata_pb.mutable_schema()));
+                mutable_new_schema, next_unique_id, col_idx_to_unique_id, tablet_metadata_pb.mutable_schema(),
+                req.__isset.compression_type ? req.compression_type : TCompressionType::LZ4_FRAME));
```
Why directly code the default compression type here?
Why not use a variable to store the default compression type.

jaogoy commented on be/src/storage/metadata_util.cpp:
```diff
@@ -225,6 +225,22 @@ Status convert_t_schema_to_pb_schema(const TTabletSchema& tablet_schema, uint32_
 
+    switch (compression_type) {
+    case TCompressionType::LZ4_FRAME:
+    case TCompressionType::LZ4:
```
Why not use a map to store the supported compression types and corresponding actions
